unity/nakama-unity-showreel/Assets/Util/NakamaManager.cs

```
@@ -0,0 +1,171 @@
+					});
+				});
+			};
+
+			_client.OnDisconnect = evt =>
+			{
+				Logger.Log("Disconnected from server.");
+				if (reconnectCount < MaxReconnectAttempts)
+				{
+					reconnectCount++;
+					_dispatchQueue.Enqueue(() => { Reconnect(); });
+				}
+				else
+				{
+					_dispatchQueue.Clear();
+					_dispatchQueue.Enqueue(() => { AfterDisconnected(this, EventArgs.Empty); });
+					
+				}
+			};
+			_client.OnError = error => ErrorHandler(error);
+		}
+
+		private IEnumerator Reconnect()
+		{
+			// if it's the first time disconnected, then attempt to reconnect immediately
+			// every other time, wait 10,20,30,40,50 seconds each time 
+			var reconnectTime = ((reconnectCount-1) + 10) * 60;  
+			yield return new WaitForSeconds(reconnectTime);
+			_sessionHandler(Session);
+		}
+		
+		private void Update()
+		{
+			for (int i = 0, l = _dispatchQueue.Count; i < l; i++)
+			{
+				_dispatchQueue.Dequeue()();
+			}
+		}
+		
+		private void OnApplicationQuit()
+		{
+			_client.Disconnect();
+		}
+		
+		private void OnApplicationPause(bool isPaused)
+		{
+			if (isPaused)
+			{
+				_client.Disconnect();
+				return;
+			}
+
+			// let's re-authenticate (if neccessary) and reconnect to the server.
+			if (_authenticateMessage != null)
+			{
+				Connect(_authenticateMessage);
+			}
+		}
+		
+		// This method connects the client to the server and
+		// if neccessary authenticates with the server
+		public void Connect(INAuthenticateMessage request)
+		{	
+			// Check to see if we have a valid session token we can restore
+			var session = RestoreSession();
+			if (session != null)
+			{
+				// Session is valid, let's connect.
+				_sessionHandler(session);
+				return;
+			}
+			
+			// Cache message for later use for reconnecting purposes in case the session had expired
+			_authenticateMessage = request;
+			
+			// Let's login to authenticate and get a valid token
+			_client.Login(request, _sessionHandler, err =>
+			{
+				ErrorHandler(err);
```

I think call this in an "else" block because we only want to report the error if it's not "ErrorCode.UserNotFound".